File: language_detail/function/recursion.c
#include <stdio.h>

// gcc -g -O2 -o recursion recursion.c
// objdump -d -M intel recursion | less
// 观察到factorial的汇编是跳转循环指令
__attribute__((noinline)) int factorial(int n, int total)
{
    if (n == 1)
    {
        return total;
    }
    return factorial(n - 1, n * total);
}

__attribute__((noinline)) int main()
{
    printf("%d\n", factorial(5, 1));
    return 0;
}

File: language_detail/function/recursion.go
package main

// 反汇编后该函数内直接调用了CALL main.factorial(SB)说明没有进行尾递归优化
func factorial(n, total int) int {
	if n == 1 {
		return total
	}
	return factorial(n-1, n*total)
}

func main() {
	println(factorial(5, 1))
}
